File: solutions/3134. Find the Median of the Uniqueness Array/3134.py
# Time complexity: O(nlogn)
# Space complexity: O(n)
class Solution:
  def medianOfUniquenessArray(self, nums: List[int]) -> int:
    n = len(nums)
    totalSubarrays = n * (n + 1) // 2
    left, right = 1, n

    while left < right:
      mid = left + (right - left) // 2
      if self.countAtLeastKdistinct(nums, mid) * 2 >= totalSubarrays:
        right = mid
      else:
        left = mid + 1

    return left

  def countAtLeastKdistinct(self, nums: List[int], k: int) -> int:
    count = 0
    left = 0
    freq = defaultdict(int)

    for right in range(len(nums)):
      freq[nums[right]] += 1

      while len(freq) > k:
        freq[nums[left]] -= 1
        if freq[nums[left]] == 0:
          del freq[nums[left]]
        left += 1

      count += right - left + 1

    return count

File: solutions/3134. Find the Median of the Uniqueness Array/3134.cpp
// Time complexity: O(nlogn)
// Space complexity: O(n)
class Solution {
 public:
  int medianOfUniquenessArray(const vector<int>& nums) {
    int n = nums.size();
    long long totalSubarrays = static_cast<long long>(n) * (n + 1) / 2;
    int left = 1;
    int right = n;

    while (left < right) {
      int mid = left + (right - left) / 2;
      if (countAtLeastKDistinct(nums, mid) * 2 >= totalSubarrays)
        right = mid;
      else
        left = mid + 1;
    }

    return left;
  }
 private:
  long long countAtLeastKDistinct(const vector<int>& nums, int k) {
    long long count = 0;
    int left = 0;
    unordered_map<int, int> freq;

    for (int right = 0; right < nums.size(); ++right) {
      ++freq[nums[right]];

      while (freq.size() > k) {
        freq[nums[left]]--;
        if (freq[nums[left]] == 0)
          freq.erase(nums[left]);
        ++left;
      }

      count += right - left + 1;
    }

    return count;
  }
};